### ISSUE TITLE
Can't switch target on v1.9.0 (VSC-1537)

### ISSUE THREAD

**Issue by michaelboeding** [espressif/vscode-esp-idf-extension#1363]
### OS

MacOS

### Operating System version

15.1.1 (24B2091)

### Visual Studio Code version

1.9.0

### ESP-IDF version

5.3.1

### Python version

3.8.7

### Doctor command output

NA

### Description

Project will not switch back to esp32s3 after going to esp32 target. Downgraded to 1.8.1 and it worked. 

### Debug Message

NA

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

**Comment by brianignacio5** [comment]
How are you switching target? What steps are you doing ? If you select a new target from the status bar it should run the `idf.py set-target esp32s3` on the background and the target shown would be updated in sdkconfig and the status bar.

If an error happened we need those doctor command output as requested in the issue template.

**Comment by atlex92** [comment]
Can confirm this bug. If I run set target via button in status bar, let's say idf.py set-target esp32s3, then id does reconfigure and update sdkconfig for esp32s3, BUT env IDF_TARGET is not updated, and if I try to build the project, it will be written that esp32 is sert as IDF_TARGET, whille project is configured for esp32s3, this is a conflict, workaround is just to manually export IDF_TARGET=esp32s3 and then continue with idf.py build

**Comment by john30** [comment]
same here, no way to change the target any more with that release, neither in idf terminal with "idf.py set-target ..." nor with the bottom bar.
misleading error reported from implicitly executed fullclean (does not matter if it was fullcleaned before or not) and the target stays on the default esp32

**Comment by brianignacio5** [comment]
Seems to me that the issue is that `I think because the bottom bar never changed` which means for some reason idf.py set-target is failing in your setup. IDF_TARGET is set in the extension build command by reading it from the sdkconfig file CONFIG_IDF_TARGET value.

Could you share the `ESP-IDF: Doctor Command`output ? I need to understand what is the error in your setup because I can't reproduce it myself just from your comments.

1. I open a blink example from ESP-IDF v5.3.1 which was originally set to esp32.
2. I click the status bar `esp32` and select `esp32s3`
3. I build the project.
4. I can see that build/project_description has `target`: esp32s3 with toolchain and others pointing to esp32s3.

@atlex92 @john30 Please also share the troubleshooting information, as a I said, I can't reproduce the issue myself but it seems that `idf.py set-target` itself is failing and that's why is not changing the IDF_TARGET.

**Comment by brianignacio5** [comment]
Do you try to build the project from IDF Terminal that you opened before changing the target ? The terminal IDF_TARGET would not be updated. Need to close and open a new terminal for IDF_TARGET changes to be seen.

**Comment by john30** [comment]
thanks for looking into it @brianignacio5 . I think it is related to https://github.com/espressif/vscode-esp-idf-extension/pull/1263 but didn't have time to look at it in depth.
just found out along the way to get it working again from a clean checkout of my project (without any vscode settings.json) that after setting idf.espIdfPath and idf.toolsPath on user level to the exact locations it works with 1.9.0 as well. proably because my env does not contain either IDF_PATH or IDF_TOOLS_PATH.
please note that i work on Windows with vscode server in WSL

**Comment by brianignacio5** [comment]
#1263 does indeed remove the idf.adapterTargetName that was used before and instead reads IDF_TARGET from sdkconfig. We have tested that it should work the same for extension commands. I haven't been able to reproduce the issue which makes me believe there is something specific happening in your setup, that is why I need to check the troubleshooting information as requested in the issue template.

**Comment by MBWeno** [comment]
My sdkconfig is not in the root of the workspace directory. When I put a dummy sdkconfig in the root with the single line containing the config_id_target set to the target I want it works.

**Comment by brianignacio5** [comment]
You could also try using the `idf.sdkconfigFilePath` in your `<project-dir>/.vscode/settings.json` to define the sdkconfig file path.

**Comment by MBWeno** [comment]
That also is working.
 
 
  On Wed, Dec 18, 2024 at 8:26 PM, Brian A. ***@***.***> wrote:   


You could also try using the idf.sdkconfigFilePath in your <project-dir>/.vscode/settings.json to define the sdkconfig file path. @MBWeno

**Comment by john30** [comment]
isn't this a chicken-egg issue? my impression when trying to get it working was that the target is reset to the default of esp32 whenever trying to change it via "idf.py set-target" or via the bottom bar action item. as it was impossible to be saved to a decent sdkconfig due to the errors reported, it wasn't able to save at all.

**Comment by brianignacio5** [comment]
> the target is reset to the default of esp32 whenever trying to change it via "idf.py set-target" or via the bottom bar action item.

This is not true. This command should set the desired IDF_TARGET to  currently defined sdkconfig file path.

The `idf.adapterTargetName` configuration setting was removed because it is duplicated, you can get it from sdkconfig file. If no sdkconfig file is defined, the default target is `esp32`.

If you don't provide any troubleshooting information, I can't see why running the idf.py set-target via bottom bar doesn't work. Alternatively, could you share an example project where this behavior is shown ? I can't reproduce the issue locally.

**Comment by brianignacio5** [comment]
No worries @michaelboeding whenever you can. Hope to fix this asap

**Comment by john30** [comment]
I've tried to get the issue on one of the example projets of esp-idf and it is easily reproducable whenever something in the build system is wrong, i.e. a part of the set-target fails. I've tested it with an invalid CMakeLists.txt in the main component (and I'm sure this can also be done for any other subcomponent) by using "INCLUDES_DIR" instead of the correct "INCLUDE_DIR".
With that precondition it is impossible to the set target. Plus, the error output gives no hint to where the problem comes from, as it tries a fullclean several times which always fails. consequently, there is no generated sdkconfig and the target stays at esp32

**Comment by brianignacio5** [comment]
@john30 Without the `ESP-IDF: Doctor Command` output, steps to reproduce and a example project you are using I can't reproduce the issue. Please share this information so I can work on a bug fix.

**Comment by john30** [comment]
please find attached the report 
[report.txt](https://github.com/user-attachments/files/18206860/report.txt)

**Comment by brianignacio5** [comment]
Well, the thing you are testing is literally the reason why the target is not updated. Because there is an issue in the build system the set-target command is not going to work, (set-target command uses the build system to generate sdkconfig) therefore sdkconfig is not updated and the target remains as `esp32`.

It seems that what you want is to override the IDF_TARGET regardless of using `set-target` command. For that case you can use the `idf.cmakeCompilerArgs` configuration setting and add IDF_TARGET there like this:

```json
"idf.cmakeCompilerArgs": [
    "-G",
    "Ninja",
    "-DPYTHON_DEPS_CHECKED=1",
    "-DESP_PLATFORM=1",
    "-DIDF_TARGET=esp32"
  ]
```

But if I add the previous setting, the build command will return something like this:

```
 *  Executing task: /usr/local/bin/cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -DIDF_TARGET=esp32 -B /Users/brian/workspace/test/blink/build -S /Users/brian/workspace/test/blink -DSDKCONFIG=/Users/brian/workspace/test/blink/sdkconfig 

CMake Error at /Users/brian/.espressif/v5.3.2/esp-idf/tools/cmake/targets.cmake:97 (message):
   IDF_TARGET 'esp32' in CMake cache does not match currently selected IDF_TARGET 'esp32c3'. To change the target, clear the build directory and sdkconfig file, and build the project again.
Call Stack (most recent call first):
  /Users/brian/.espressif/v5.3.2/esp-idf/tools/cmake/project.cmake:24 (__target_init)
  CMakeLists.txt:5 (include)
```

This is because the sdkconfig file and the build directory both need to be deleted before running the build command with `-DIDF_TARGET-esp32`. Now if you delete them, the build command will work.

I hope this output explain why a successful set-target command is necessary to update the IDF_TARGET, regardless of a configuration setting that force a new target in the build system.

**Comment by larryk3** [comment]
I am also experiencing odd issues with switching the target. I have set my `esp_idf_project_configuration.json` to use a custom `sdkconfigDefaults`, `sdkconfigFilePath`, and `buildDirectoryPath`.

My configuration can be seen on this branch in my project - https://github.com/K3Engineering/dynamite-sampler-firmware/tree/add-multiple-configurations

I did the following:
- Start with a clean repo
- Change the esp target from `esp32` to `esp32s3` by clicking the button at the bottom of vscode
- Open a terminal and run `$Env:IDF_TARGET` (this prints the env variable in powershell) - the result is `esp32s3`
- Delete the newly created build directory (in my case it is `build.debug`, as defined in my `esp_idf_project_configuration.json`)
- Open a terminal and run `$Env:IDF_TARGET` - the result is now `esp32`

After this point, if I try to build, it will build a `esp32` target. And I have to switch targets again to make it work. So after every clean, my `IDF_TARGET` env variable is set back to `esp32`.

This was done on plugin version `1.9.0`. I tried the same thing with `1.8.1` and it didn't have this bug! After cleaning the build directory and getting the `IDF_TARGET`, it would properly show `esp32s3`.

**Comment by brianignacio5** [comment]
Does your `sdkconfigFilePath` is defined inside the build directory ?

**Comment by brianignacio5** [comment]
I think I come up with a fix that makes sense  in #1385 by saving IDF_TARGET in `idf.customExtraVars` and use it when the value from sdkconfig if not available.

Please try this [vsix installer](https://nightly.link/espressif/vscode-esp-idf-extension/actions/artifacts/2362593445.zip).

You can test these changes by installing this VSIX by click menu View -> Command Palette..., type Install from VSIX and then select downloaded esp-idf-extension.vsix file to install the extension.
